proposals/xxxx-nested-protocols.md

```diff
@@ -0,0 +1,150 @@
+* Upcoming Feature Flag: `NestedProtocols`
+* Review: ([pitch](https://forums.swift.org/...))
+
+## Introduction
+
+Allows protocols to be nested in non-generic `struct/class/enum/actor`s, and functions.
+
+## Motivation
+
+Nesting nominal types inside other nominal types allows developers to express a natural scope for the inner type -- for example, `String.UTF8View` is `struct UTF8View` nested within `struct String`, and its name clearly communicates its purpose as an interface to the UTF-8 code-units of a String value.
+
+However, nesting is currently restricted to struct/class/enum/actors within other struct/class/enum/actors; protocols cannot be nested at all, and so must always be top-level types within a module. This is unfortunate, and we should relax this restriction so that developers can express protocols which are naturally scoped to some outer type.
+
+## Proposed solution
+
+We would allow nesting protocols within non-generic struct/class/enum/actors, and also within functions that do not belong to a generic context.
+
+For example, `TableView.Delegate` is naturally a delegate protocol pertaining to table-views. Developers should be declare it as such - nested within their `TableView` class:
+
+```swift
+class TableView {
+  protocol Delegate: AnyObject {
+    func tableView(_: TableView, didSelectRowAtIndex: Int)
+  }
+}
+
+class DelegateConformer: TableView.Delegate {
+  func tableView(_: TableView, didSelectRowAtIndex: Int) {
+    // ...
+  }
+}
+```
+
+Currently, developers resort to giving things compound names, such as `TableViewDelegate`, to express the same natural scoping that could otherwise be expressed via nesting.
+
+As an additional benefit, within the context of `TableView`, the nested protocol `Delegate` can be referred to by a shorter name (as is the case with all other nested types):
+
+```swift
+class TableView {
+  weak var delegate: Delegate?
+  
+  protocol Delegate { /* ... */ }
+}
+```
+
+Protocols can also be nested within non-generic functions and closures. Admittedly, this is of somewhat limited utility, as all conformances to such protocols must also be within the same function. However, there is also no reason to artificially limit the complexity of the models which developers create within a function. Some codebases (of note, the Swift compiler itself) make use of large closures with nested types, and they beneift from abstractions using protocols.
+
+```swift
+func doSomething() {
+
+   protocol Abstraction {
+     associatedtype ResultType
+     func requirement() -> ResultType
+   }
+   struct SomeConformance: Abstraction {
+     func requirement() -> Int { ... }
+   }
+   struct AnotherConformance: Abstraction {
+     func requirement() -> String { ... }
+   }
+   
+   func impl<T: Abstraction>(_ input: T) -> T.ResultType {
+     // ...
+   }
+   
+   let _: Int = impl(SomeConformance())
+   let _: String = impl(AnotherConformance())
+}
+```
+
+## Detailed design
```

Comment:
A question came up in the pitch thread about whether or not a nested protocol satisfies an associated type requirement (as an existential type) of the same name in the same way that nested types or typealiases do: https://forums.swift.org/t/pitch-allow-protocols-to-be-nested-in-non-generic-contexts/65285/22. It's worth specifying the witness matching behavior in the detailed design section here.